### PR TITLE
[NEW] document_type module

### PR DESCRIPTION
**Pull request by LeartS**
Define type of documents per model, and choose what document you are uploading
when attaching a file to a record.

* Go to Knowledge / Configuration / Document types and configure your document types
* Every time you attach a document using the 'Attachment' menu in the sidebar,
  a popup will open allowing you to set attachment name, description and type;
  according to the types defined for that model.
* In the document list, you can filter and/or groupby by document type.

### CODE REVIEW

File: document_type/README.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+* Go to Knowledge / Configuration / Document types and configure your document types
+* Every time you attach a document using the 'Attachment' menu in the sidebar,
+  a popup will open allowing you to set attachment name, description and type;
+  according to the types defined for that model.
+
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/repo/118/9.0
```
Comment (max3903) [inline document_type/README.rst]: Should be `https://runbot.odoo-community.org/runbot/repo/118/10.0`

**Review by max3903 — CHANGES_REQUESTED** [review]

File: document_type/README.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/
+knowledge/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback `here <https://github.com/OCA/
```
Comment (elicoidal) [inline document_type/README.rst:35]: Use the newest template

File: document_type/README.rst [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+Acknowledgements
+----------------
+
+* Thanks to gingitsune for showing me a simple way to open a 'Save and Close' dialog
+
+Maintainer
+----------
+
+.. image:: https://odoo-community.org/logo.png
```
Comment (elicoidal) [inline document_type/README.rst:62]: please newest template (https)

File: document_type/models/document_type.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from odoo import fields, models
```
Comment (elicoidal) [inline document_type/models/document_type.py:4]: Please add the copyright short headers in all py, xml and js files

**Review by elicoidal — CHANGES_REQUESTED** [review]
Some details. Not tested

File: document_type/README.rst (hunk 1 above)
Comment (dreispt) [reply document_type/README.rst]: +1

File: document_type/models/document_type.py (hunk 4 above)
Comment (dreispt) [reply document_type/models/document_type.py:4]: +1

**Review by dreispt — COMMENTED** [review]
Build is failing on the `group_document_manager` creation. Can you check?

**Comment by LeartS** [comment]
Rebased, fixed README link, fixed manifest, added copyright notices. @elicoidal @max3903 @dreispt

**Review by max3903 — APPROVED** [review]

File: document_type/models/document_type.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    _description = 'Document Type'
+
+    name = fields.Char(required=True)
+    ir_model_id = fields.Many2one('ir.model', string='Object')
+    document_ids = fields.One2many(
+        'ir.attachment', 'document_type_id', string='Documents')
+    documents_count = fields.Integer(compute='_compute_documents_count')
+
+    def _compute_documents_count(self):
+        for record in self:
```
Comment (lmignon) [inline document_type/models/document_type.py:18]: For performance it could be more efficient to use `read_group`

**Review by lmignon — CHANGES_REQUESTED** [review]
A document type can be linked to 0 or 1 model. Nevertheless it seems that a document type not linked to a specific model can't be used.... I expect that if no model is specified on the document type, this type can be used everywhere. 
It could be nice to also allow to use the same document type on a list of models. (many2many)
@max3903 @elicoidal If you think it's not blocking, I don't block the merge...

**Comment by LeartS** [comment]
Both good points, I'll implement them

**Pull request closed without merging**